### ISSUE TITLE
Work on performance when integrating results from shards on the Coordinator

### ISSUE THREAD

**Issue by premvvnc** [arangodb/arangodb#3449]
## my environment running ArangoDB
I'm using the latest ArangoDB of the respective release series:
- [ ] 2.8
- [ ] 3.0
- [ ] 3.1
- [x] 3.2
- [ ] self-compiled devel branch

Mode:
- [x] Cluster
- [ ] Single-Server

Storage-Engine:
- [x] mmfiles
- [ ] rocksdb

On this operating system:
- [ ] DCOS on
  - [x] AWS
  - [ ] Azure
  - [ ] own infrastructure
- [ ] Linux 
  - [ ] Debian .deb
  - [x] Ubuntu .deb
  - [ ] SUSE   .rpm
  - [ ] RedHat .rpm
  - [ ] Fedora .rpm
  - [ ] Gentoo
  - [ ] docker - official docker library
  - [ ] other:
- [ ] Windows, version: 
- [ ] MacOS, version:

### this is an AQL-related issue:
[ ] I'm using graph features

I'm issuing AQL via:
- [x] web interface with this browser:     running on this OS:
- [ ] arangosh
- [ ] this Driver:

I've run `db._explain("<my aql query>")` and it didn't shed more light on this.
The AQL query in question is:

The issue can be reproduced using this dataset:

Please provide a way to create the dataset to run the above query on; either by a gist with an arangodump, or `db.collection.save({my: "values"}) statements. If it can be reproduced with one of the ArangoDB example datasets, it's a plus.

### this is a web interface-related issue:
I'm using the web interface with this browser:     running on this OS:
- [x] authentication is enabled?
- [x] using the cluster?
- [ ] _system database?

These are the steps to reproduce:
1) open the browser on http://127.0.0.1:8529
2) log in as ...
3) use database [ ] `_system` [ ] other: 
4) click to ...
...

The following problem occurs: [Screenshot?] 
Instead I would be expecting: 

### this is an installation-related issue:
Describe which steps you carried out, what you expected to happen and what actually happened.

**Comment by premvvnc** [comment]
Hi we have have set up a 3 node memory mapped arangodb and created a collection with 9 shards and inserted 400k documents, Ran a query with filter on date like  > '2016-01-01T.....'  and the time it took is about 500ms, so tried to reduce the query time by adding a skiplist index on date and when I ran the query it took like 1.5 secs which is very strange, expected indexes to speed up the query but this is quite opposite,Played with index bucket sizes but did not help. Has anyone faced similar problem.

Query string:
 LET queryOne = ( for doc in test
 FILTER doc.date > "2014-01-01T00:00:00+00:00" 
 COLLECT WITH COUNT INTO length return length ) 
 RETURN { ac: queryOne }

Execution plan:
 Id   NodeType          Site         Est.   Comment
  1   SingletonNode     COOR            1   * ROOT
  8   SubqueryNode      COOR            1     - LET queryOne = ...   /* const subquery */
  2   SingletonNode     DBS             1       * ROOT
 11   IndexNode         DBS        199174         - FOR doc IN test   /* skiplist index scan */
 14   RemoteNode        COOR       199174           - REMOTE
 15   GatherNode        COOR       199174           - GATHER doc ASC
  6   CollectNode       COOR            1           - COLLECT  WITH COUNT INTO length   /* sorted */
  7   ReturnNode        COOR            1           - RETURN length
  9   CalculationNode   COOR            1     - LET #7 = { "ac" : queryOne }   /* simple expression */
 10   ReturnNode        COOR            1     - RETURN #7

Indexes used:
 By   Type       Collection              Unique   Sparse   Selectivity   Fields                    Ranges
 11   skiplist   test   false    false            n/a   [ `date` ]   (doc.`date` > "2014-01-01T00:00:00+00:00")

Optimization rules applied:
 Id   RuleName
  1   use-indexes
  2   remove-filter-covered-by-index
  3   remove-unnecessary-calculations-2
  4   scatter-in-cluster
  5   remove-unnecessary-remote-scatter

**Comment by ObiWahn** [comment]
TestCase

```arangosh
db._create("test", {"replicationFactor":3, "numberOfShards": 9})
```

```bash
main(){
    local start="$1"
    local end="$2"
    local count="$3"

    echo "creating $count dates between $start and $end" >&2

    start_s=$(date -d"$start" +%s)
    end_s=$(date -d"$end" +%s)
    diff_s=$((end_s - start_s))

    for i in $(seq 1 $count); do
        local rand=$(( ((RANDOM<<15)|RANDOM) % diff_s ))
        local d=$((start_s + rand))
        date --date="@$d" +%F-%H:%M:%S
    done
}

main "$@"
```

```sh
for date in $(./test/generate_dates.sh 1999-01-01 2020-01-01 400000); do curl -L -XPOST http://[::1]:9196/_api/document/ulf -d "{ \"date\":\"$date\" }"; done;
```

**Comment by ObiWahn** [comment]
the lines below confirm the behavior:
when the index filters much then the query becomes faster than without index
```
[::1]:9196@_system> q='LET queryOne = ( for doc in ulf FILTER doc.date > "2014-01-01T00:00:00+00:00" COLLECT WITH COUNT INTO length return length ) RETURN { ac: queryOne }'
LET queryOne = ( for doc in ulf FILTER doc.date > "2014-01-01T00:00:00+00:00" COLLECT WITH COUNT INTO length return length ) RETURN { ac: queryOne }

[::1]:9196@_system> db._query(q).next().ac[0]
70629

[::1]:9196@_system> rv=0; for(var i=0; i<100; i++){ var start=Date.now(); db._query(q); rv+=(Date.now() - start);}; print(rv/100 + "ms");
483.59ms

[::1]:9196@_system> db._explain(q)
Query string:
 LET queryOne = ( for doc in ulf FILTER doc.date > "2014-01-01T00:00:00+00:00" COLLECT WITH COUNT INTO length return length ) RETURN { ac: queryOne }
Execution plan:
 Id   NodeType                  Site         Est.   Comment
  1   SingletonNode             COOR            1   * ROOT
  8   SubqueryNode              COOR            1     - LET queryOne = ...   /* const subquery */
  2   SingletonNode             DBS             1       * ROOT
  3   EnumerateCollectionNode   DBS        400000         - FOR doc IN ulf   /* full collection scan */
  4   CalculationNode           DBS        400000           - LET #5 = (doc.`date` > "2014-01-01T00:00:00+00:00")   /* simple expression */   /* collections used: doc : ulf */
  5   FilterNode                DBS        400000           - FILTER #5
 13   RemoteNode                COOR       400000           - REMOTE
 14   GatherNode                COOR       400000           - GATHER 
  6   CollectNode               COOR            1           - COLLECT  WITH COUNT INTO length   /* sorted */
  7   ReturnNode                COOR            1           - RETURN length
  9   CalculationNode           COOR            1     - LET #7 = { "ac" : queryOne }   /* simple expression */
 10   ReturnNode                COOR            1     - RETURN #7
Indexes used:
 none
Optimization rules applied:
 Id   RuleName
  1   scatter-in-cluster
  2   distribute-filtercalc-to-cluster
  3   remove-unnecessary-remote-scatter

[::1]:9196@_system> db.ulf.ensureIndex({ type:"skiplist", fields:["date"]})
{ 
  "deduplicate" : true, 
  "fields" : [ 
    "date" 
  ], 
  "id" : "ulf/416854", 
  "isNewlyCreated" : true, 
  "sparse" : false, 
  "type" : "skiplist", 
  "unique" : false, 
  "code" : 201 
}

[::1]:9196@_system> db._query(q).next().ac[0]
70629

[::1]:9196@_system> rv=0; for(var i=0; i<100; i++){ var start=Date.now(); db._query(q); rv+=(Date.now() - start);}; print(rv/100 + "ms");
729.49ms

[::1]:9196@_system> db._explain(q)
Query string:
 LET queryOne = ( for doc in ulf FILTER doc.date > "2014-01-01T00:00:00+00:00" COLLECT WITH COUNT INTO length return length ) RETURN { ac: queryOne }
Execution plan:
 Id   NodeType          Site         Est.   Comment
  1   SingletonNode     COOR            1   * ROOT
  8   SubqueryNode      COOR            1     - LET queryOne = ...   /* const subquery */
  2   SingletonNode     DBS             1       * ROOT
 11   IndexNode         DBS        200000         - FOR doc IN ulf   /* skiplist index scan */
 14   RemoteNode        COOR       200000           - REMOTE
 15   GatherNode        COOR       200000           - GATHER doc ASC
  6   CollectNode       COOR            1           - COLLECT  WITH COUNT INTO length   /* sorted */
  7   ReturnNode        COOR            1           - RETURN length
  9   CalculationNode   COOR            1     - LET #7 = { "ac" : queryOne }   /* simple expression */
 10   ReturnNode        COOR            1     - RETURN #7
Indexes used:
 By   Type       Collection   Unique   Sparse   Selectivity   Fields       Ranges
 11   skiplist   ulf          false    false            n/a   [ `date` ]   (doc.`date` > "2014-01-01T00:00:00+00:00")
Optimization rules applied:
 Id   RuleName
  1   use-indexes
  2   remove-filter-covered-by-index
  3   remove-unnecessary-calculations-2
  4   scatter-in-cluster
  5   remove-unnecessary-remote-scatter

**Comment by ObiWahn** [comment]
With Index we have non simple GatherBlocks (doing merge sort). The sort takes a lot more time and there is more http traffic involved than in the simple case. We are trying to improve the sort algorithm and we will check if there is a way to avoid merge sorts if the index is used for filtering and no sort is required.

**Comment by ObiWahn** [comment]
The error is triggerd in this code:

arangod/RocksDBEngine/RocksDBGeoIndexImpl.cpp
``` c++
354     inline void RocksRead(GeoIx* gix, RocksDBKey const& key, std::string* val) {
355       arangodb::Result r =
356           gix->rocksMethods->Get(RocksDBColumnFamily::geo(), key, val);
357       if (!r.ok()) {
358         LOG_TOPIC(ERR, Logger::FIXME) << r.errorMessage();
359         THROW_ARANGO_EXCEPTION_MESSAGE(r.errorNumber(), r.errorMessage());
360       }
361     }
```

**Comment by ObiWahn** [comment]
Another failure triggered by the test in the rocksdb geoindex
```
(gdb) bt
#0  0x00007ffff62a9428 in __GI_raise (sig=sig@entry=6) at ../sysdeps/unix/sysv/linux/raise.c:54
#1  0x00007ffff62ab02a in __GI_abort () at abort.c:89
#2  0x00007ffff62eb7ea in __libc_message (do_abort=do_abort@entry=1, fmt=fmt@entry=0x7ffff640345f "*** %s ***: %s terminated\n") at ../sysdeps/posix/libc_fatal.c:175
#3  0x00007ffff638d11c in __GI___fortify_fail (msg=<optimized out>, msg@entry=0x7ffff6403441 "stack smashing detected") at fortify_fail.c:37
#4  0x00007ffff638d0c0 in __stack_chk_fail () at stack_chk_fail.c:28
#5  0x00000000023ad953 in arangodb::rocksdbengine::GeoIndex_insert (gi=0x7fffe5161500, c=0x7fffd02471f0) at /mnt/c9/jan/arangodb4/arangod/RocksDBEngine/RocksDBGeoIndexImpl.cpp:1845
```

**Comment by dothebart** [comment]
@ObiWahn s work ( 987aea2af34c6d018797a37d304159fd1c66a378 ) will be part of the upcomming 3.3 release. 
There will be more work on this issue which I will reflect in the title.

**Comment by maxkernbach** [comment]
This issue has been addressed in 3.3. 

Heap merge for cluster blocks, which improves the performance in a cluster significantly, was implemented in the following pull request: https://github.com/arangodb/arangodb/pull/3617
